src/org/javarosa/form/api/FormEntryController.java

```diff
@@ -72,12 +72,13 @@ public int answerQuestion(IAnswerData data, boolean midSurvey) {
 
 
     /**
-     * Attempts to save the answer at the specified FormIndex into the
-     * datamodel.
+     * Attempts to save the answer at the specified FormIndex into the datamodel.
```

Comment:
I'm missing information about the return values here. The explanation on the `@param` can be missed out and we could give a bit more information.

It doesn't explain when the side effects happen (they don't always take place), and there are another two undocumented side effects (when the question is complex, and the `RuntimeException`) that should be included:
```java
/**
 * Attempts to save the answer at the given {@link FormIndex} into the instance
 * and returns one of three possible {@code int} attempt result codes:
 * <ul>
 * <li>{@link #ANSWER_OK}
 * <li>{@link #ANSWER_REQUIRED_BUT_EMPTY}
 * <li>{@link #ANSWER_CONSTRAINT_VIOLATED}
 * </ul>
 * <p>
 * Side effects: When it returns {@link #ANSWER_OK}, it mutates
 * the {@link TreeElement} corresponding to the given {@link FormIndex} by
 * setting its value to the given {@link IAnswerData} or by copying an
 * itemset answer if the question is complex.
 *
 * @param index The index of the question/prompt that is being currently evaluated
 * @param data  The data to attempt to answer the question with.
 * @return the attempt's {@code int} result code
 * @throws RuntimeException when the question is complex and it has constraints.
 *                          See inline comments.
 * @see QuestionDef#isComplex()
 */
```
(I've also changed `dataModel` with `instance`, which is more consistently used to refer to the "tree of answers filled by the user")

Reply:
This is way better. Thanks particularly for your attention to side effects. I have updated in a new commit.